### PR TITLE
Ports balloon alerts

### PR DESCRIPTION
**Pull request by Archanial**
## About The Pull Request

Ports balloon alerts from tg
Ported prs:
https://github.com/tgstation/tgstation/pull/59130
https://github.com/tgstation/tgstation/pull/59290
https://github.com/tgstation/tgstation/pull/59256
https://github.com/tgstation/tgstation/pull/59255
some code from https://github.com/tgstation/tgstation/pull/59232

Guns and energy sword use balloon alerts too.
balloon_alert

Storage uses balloon alerts.
Reagent containers also use balloon.
Walls and catwalks too.

I've made balloon alert a subtype of chatmessage so it is handled by SSrunechat.
Main difference between it and actual message is that it is not added to `seen_messages` thus not counted while changing position of previous message.

- [x] Make balloon alerts use SSrunechat
- [x] Walls
- [x] Tools
- [x] Better CL

## Why It's Good For The Game

Looks nice and is used by wiremod

## Changelog
:cl: Mothblocks,  dragomagol, InsaneRed, Archanial
add: Balloon alerts - small runechat like pop ups while performing specific actions, which can be toggled in preferences
add: Balloon alerts to some mech actions, door remote and beakers
add: Balloon alerts to storage, guns, energy swords, reagent containers, constructing walls and tools
/:cl:

### CODE REVIEW

**Review by PowerfulBacon — COMMENTED** [review]
I am going to die from the punctuation inconsistencies

File: code/game/mecha/mecha_actions.dm [hunk 1]
```diff
@@ -176,12 +176,12 @@
...
 		chassis.step_in = min(1, round(chassis.step_in/2))
 		chassis.step_energy_drain = max(chassis.overload_step_energy_drain_min,chassis.step_energy_drain*chassis.leg_overload_coeff)
-		chassis.occupant_message("<span class='danger'>You enable leg actuators overload.</span>")
+		chassis.balloon_alert(owner,"Leg actuators overloaded")
 	else
 		chassis.leg_overload_mode = 0
 		chassis.step_in = initial(chassis.step_in)
 		chassis.step_energy_drain = chassis.normal_step_energy_drain
-		chassis.occupant_message("<span class='notice'>You disable leg actuators overload.</span>")
+		chassis.balloon_alert(owner, "You disable the overload")
```
Comment (PowerfulBacon) [inline code/game/mecha/mecha_actions.dm]:
```suggestion
		chassis.balloon_alert(owner, "Leg actuators reset")
```

File: code/game/objects/items/control_wand.dm [hunk 2]
```diff
@@ -28,7 +28,7 @@
 			mode = WAND_EMERGENCY
 		if(WAND_EMERGENCY)
 			mode = WAND_OPEN
-	to_chat(user, "Now in mode: [mode].")
+	balloon_alert(user, "Mode: [mode]")
```
Comment (PowerfulBacon) [inline code/game/objects/items/control_wand.dm]: I dont like how this is the only one using : compared to all the other ones which are Mode set to [mode] or something.

Comment (Archanial) [reply code/game/objects/items/control_wand.dm]: eh
```suggestion
	balloon_alert(user, "Mode set to [mode]")
```

**Comment by Archanial** [comment]
Making this draft because it has a lot of potential I'm going to explore

**Comment by ike709** [comment]
Code's fine. Needs a better CL.

**Review by ike709 — DISMISSED** [review]
Just needs a final testmerge

File: code/datums/chatmessage.dm [hunk 3]
```diff
@@ -346,6 +353,84 @@ GLOBAL_LIST_INIT(job_colors_pastel, list(
...
+	)
+
+	animate(
+		alpha = 0,
+		time = BALLOON_TEXT_FULLY_VISIBLE_TIME,
+		easing = CUBIC_EASING | EASE_IN,
+	)
+
+	// Register with the runechat SS to handle EOL and destruction
+	scheduled_destruction = world.time + (lifespan - CHAT_MESSAGE_EOL_FADE)
```
Comment (Archanial) [inline code/datums/chatmessage.dm]:
```suggestion
	scheduled_destruction = world.time + BALLOON_TEXT_TOTAL_LIFETIME
```

File: code/modules/surgery/helpers.dm [hunk 4]
```diff
@@ -69,12 +69,14 @@
...
-				user.visible_message("[user] drapes [I] over [M]'s [parse_zone(selected_zone)] to prepare for surgery.", \
-					"<span class='notice'>You drape [I] over [M]'s [parse_zone(selected_zone)] to prepare for \an [procedure.name].</span>")
+				user.visible_message("[user] drapes [I] over [M]'s [parse_zone(selected_zone)] to prepare for surgery.",
+				"<span class='notice'>You drape [I] over [M]'s [parse_zone(selected_zone)] to prepare for \an [procedure.name].</span>")
+				I.balloon_alert(user, "You drape over [parse_zone(selected_zone)]")
 
 				log_combat(user, M, "operated on", null, "(OPERATION TYPE: [procedure.name]) (TARGET AREA: [selected_zone])")
 			else
-				to_chat(user, "<span class='warning'>You need to expose [M]'s [parse_zone(selected_zone)] first!</span>")
+				I.balloon_alert(user, "[parse_zone(selected_zone)] is covered up")
```
Comment (PowerfulBacon) [inline code/modules/surgery/helpers.dm:78]: messages like this 100% need to be sent to the chat as well

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
All cases of to_chat being removed should be readded and have both to_chat messages and balloon alert messages, or the balloon alert should automatically print to chat. If the chat messages are disabled, then the game should still have messages, plus scrolling up in chat to see what happened is important.

**Comment by ike709** [comment]
> All cases of to_chat being removed should be readded

Haha, no. What should be done is have a pref for balloon alert, chat output, or both. And just have the proc output wherever based on that.

File: code/datums/chatmessage.dm [hunk 5]
```diff
@@ -346,6 +357,102 @@ GLOBAL_LIST_INIT(job_colors_pastel, list(
 		if(5)
 			return "#[num2hex(c, 2)][num2hex(m, 2)][num2hex(x, 2)]"
 
+/atom/proc/balloon_alert(mob/viewer, text)
+	if(!viewer.client)
+		return
+	var/client/C = viewer.client
+	switch(C.prefs.see_balloon_alerts)
```
Comment (ike709) [inline code/datums/chatmessage.dm]:
```suggestion 
	switch(viewer.client.prefs.see_balloon_alerts)
```

**Review by ike709 — APPROVED** [review]
YOLO

**Pull request merged**